rust/src/quic/parser.rs

```diff
@@ -0,0 +1,330 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+use super::error::QuicError;
+use super::frames::Frame;
+use nom::bytes::complete::take;
+use nom::number::complete::{be_u32, be_u8};
+
+/*
+   gQUIC is the Google version of QUIC.
+
+   The following docs were referenced when writing this parser
+
+   References:
+       - https://docs.google.com/document/d/1WJvyZflAO2pq77yOLbp9NsGjC1CHetAXV8I0fQe-B_U/edit
+       - https://docs.google.com/document/d/1g5nIXAIkN_Y-7XJW5K45IblHd_L2f5LTaDUDwvZ5L6g/edit
+       - https://www.slideshare.net/shigeki_ohtsu/quic-overview
+       - https://tools.ietf.org/html/draft-ietf-quic-transport-17#section-19.8
+       - https://github.com/salesforce/GQUIC_Protocol_Analyzer/blob/master/src/gquic-protocol.pac
+*/
+
+// List of accepted and tested quic versions format
+#[derive(Debug, PartialEq, Clone)]
+pub enum QuicVersion {
+    Q043,
+    Q044,
+    Q045,
+    Q046,
+
+    Unknown(u32),
+}
+
+impl QuicVersion {
+    fn is_gquic(&self) -> bool {
+        *self == QuicVersion::Q043
+            || *self == QuicVersion::Q044
+            || *self == QuicVersion::Q045
+            || *self == QuicVersion::Q046
+    }
+}
+
+impl From<QuicVersion> for u32 {
+    fn from(from: QuicVersion) -> Self {
+        match from {
+            QuicVersion::Q043 => 0x51303433,
```

Comment:
I do not know if for detection, we should translate 3 -> 0x51303433...

Reply:
not sure I understand what you mean

Reply:
For the signature keyword `quic.version` do we match a packet with `QuicVersion::Q043` if the signature has `quic.version: 3` or `quic.version: 0x51303433` ?

Reply:
Ah, yes. Right now it's parsing the keyword in C using `DetectU32Parse`